JEP.md

```diff
@@ -0,0 +1,293 @@
+
+```typescript
+import {EventSink, EventData} from '@jupyterlab/telemetry'
+
+export class MyCustomEventSink implements EventSink {
+    handle_event(eventData: EventData): Promise<void> {
+        // Business logic to publish event to backend.
+    }
+}
+```
+
+#### JupyterLab Configuration
+
+The ability to configure the event sinks to publish to as well as events to blacklist or whitelist events. This can likely be accomplished via JupyterLab settings and the JupyterLab dependency management mechanism.
+
+The JupyterLab telemetry extension provides `Token<EventSink>` that event sink implementations can depend on and register themselves with the  core event router.
+
+#### JupyterLab Core event router
+
+The implementation of the `@jupyterlab/telemetry/recordEvent` method which handles routing events to the configured sinks and additionally:
+
+* Schema validation
+* Adds event metadata such as `creationTime`
+* Filters events based on the configuration
+* (Optionally) Aggregation and data cleaning.
+
+#### User opt-in
+
+Since JupyterLab is the user facing component, it also contains UX features to give more visibility and transparency to the Jupyter user. In general, every component should make it clear to the user what data is being collected and possible ways to turn it off.
+
+* UI for opting in or opting out of telemetry data collection
+* UI for showing the list of events that are currently being collected.
+
+### JupyterHub
+
+(This section needs to be filled out)
+
+### Jupyter Classic
+
+The proposal for Jupyter Classis is to have a convenience JS library that can be used to pubish events to the server [REST Endpoint](#rest-endpoint).
+This ensures that we provide support for Jupyter Classic but can rely on the Jupyter Server to do much of the heavy-lifting by relying on the Core Event Router, Event Sinks, and configuration done at the server level.
+
+```javascript
+
+var telemetry = require('jupyter-telemetry-js')
+
+telemetry.record_event(
+    name='org.jupyter.kernel_lifecycle_event',
+    schema_version='1',
+    args={
+        'kernel_name': 'python3',
+        'state': 'restarted'
+    }
+)
+```
+
+### Data protection
+
+(This section needs to be filled out)
+
+* What guidance do we provide? How do we provide this?
+* What building blocks do we provide?
+
+## Pros and Cons
+
+PROS
+
+* Implementation shares common principles and components across the Jupyter components.
+* Decoupling between event publishers and event consumers.
+* Flexible configuration for Jupyter operators to configure event sinks and event names.
+
+CONS
+
+* How do we enable data protection by default? (Just marking a event attribute as PII may not solve this)
+* What guidance and building blocks do we provide for compliance programs such as GDPR
```

Comment:
I have these as *Cons* for now since we aren't giving concrete solutions for these. I'll reword these to be statements and once we address these we can move them out of the cons.